File: internal/infra/di/generate_payment.go
package di

import (
	driven_dynamodb "github.com/fiap/ms-payment/internal/adapters/driven/dynamodb"
	driven_mercadopago "github.com/fiap/ms-payment/internal/adapters/driven/mercadopago"
	"github.com/fiap/ms-payment/internal/core/application/usecases"
	infra_aws "github.com/fiap/ms-payment/internal/infra/aws"
)

type GeneratePaymentContainer struct {
	GeneratePaymentUseCase *usecases.GeneratePaymentUseCase
}

func NewGeneratePaymentContainer(region infra_aws.AwsRegion, stage infra_aws.AwsStage) *GeneratePaymentContainer {
	dynamoClient := infra_aws.NewDynamoDBClient(region, stage)
	smClient := infra_aws.NewSMClient(region, stage)

	paymentRepo := driven_dynamodb.NewPaymentRepository(*dynamoClient)
	generateTokenMP := driven_mercadopago.NewMercadoPagoGenerateToken(smClient)
	generateQRCodeMP := driven_mercadopago.NewMercadoPagoGenerateQRCode(smClient)

	return &GeneratePaymentContainer{
		GeneratePaymentUseCase: &usecases.GeneratePaymentUseCase{
			PaymentRepo:      paymentRepo,
			GenerateTokenMP:  generateTokenMP,
			GenerateQRCodeMP: generateQRCodeMP,
		},
	}
}

File: src/infra/di/generate_payment/generate_payment.ts
import { MercadoPagoGenerateQRCode } from "@driven_mercadopago/generate_qrcode"
import { MercadoPagoGenerateToken } from "@driven_mercadopago/generate_token"
import { PaymentRepository } from "@driven_dynamodb/payment_repository"
import { GeneratePaymentUseCase } from "@usecases/generate_payment"
import { DynamoDBClientWrapper } from "@aws/dynamodb_client"
import { SNSClientWrapper } from "@aws/sns_client"
import { AwsRegion, AwsStage } from "@aws/utils"
import { SMClientWrapper } from "@aws/sm_client"
import { PaymentSNS } from "@driven_sns/payment"

export class GeneratePaymentContainerFactory {
	usecase: GeneratePaymentUseCase

	constructor(region: AwsRegion, stage: AwsStage) {
		const dynamoClient = new DynamoDBClientWrapper(region, stage)
		const smClient = new SMClientWrapper(region, stage)
		const snsClient = new SNSClientWrapper(region, stage)

		const paymentRepo = new PaymentRepository(dynamoClient)
		const generateTokenMP = new MercadoPagoGenerateToken(smClient)
		const generateQRCodeMP = new MercadoPagoGenerateQRCode(smClient)
		const paymentTopic = new PaymentSNS(snsClient)

		this.usecase = new GeneratePaymentUseCase(paymentRepo, generateTokenMP, generateQRCodeMP, paymentTopic)
	}
}
